BLD: Update to v1.49.0; use nvimgcodec package; make CUDA optional

carterbox opened the pull request:
Updates the source to v1.49.0 in order to allow use of `nvimgcodec=0.5` which is now available as a conda package instead of building from redist here.

Switches CUDA dependencies from dynamic compile time links to dynamic runtime links (dlopen) so that CUDA features are optional without needing to ship a separate build variant for CUDA vs non-CUDA. As part of this, the CUDA runtime library is statically linked, and the FFTS library is now revended and linked statically (because FFT features require a CPU fallback?). Users now need to manually opt-in to installing NVIDIA CUDA libraries such as nvimgcodec, nvjpeg, cufile, etc.

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
Closes #35

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ The build section contained an unexpected subsection name. skup is not a valid subsection name.

For **recipe/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14543685325](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14543685325). Examine the logs at this URL for more detail.</sub>

JanuszL commented on recipe/bld.bat:
```diff
@@ -0,0 +1,62 @@
+
+mkdir third_party\boost\preprocessor\include
```
DALI doesn't support Windows at the moment as there are places that are Linux-specific and won't work on Windows by definition (path separators, specific OS syscalls).

JanuszL commented on recipe/build.sh:
```diff
@@ -19,31 +19,52 @@ sed -i.bak "s/@DALI_INSTALL_REQUIRES_NVIMGCODEC@//g" dali/python/setup.py.in
+# Continue to dlopen nvimgcodec so that it can be optionally installed
+  -DWITH_DYNAMIC_NVIMGCODEC=ON
+  -DNVIMGCODEC_DEFAULT_INSTALL_PATH=${PREFIX}
+# Disable all dynamic (dlopen) linkages because we have patched the static links away
+  -DWITH_DYNAMIC_CUDA_TOOLKIT=OFF
+  -DWITH_DYNAMIC_CUFFT=OFF
+  -DWITH_DYNAMIC_NPP=OFF
+  -DWITH_DYNAMIC_NVJPEG=OFF
```
The preference is to link dynamically to shrink the size of the binary.

carterbox replied:
As noted in the comment. We have patched the static links away. These libraries are linked dynamically, just not with dlopen.

JanuszL replied:
Any particular reason to move away from dlopen in favor of this approach which diverges from the upstream code?

carterbox replied:
Lets us check at build time if runtime dependencies are satisfied by inspecting the binary elf record for links and undefined symbols. The build runners don't have GPUs so we can't actually load the binaries to see if they fail because trying to load the cuda driver will always fail.

What's the reason for using dlopen?

carterbox replied:
Making CUDA optional at runtime with a single build variant is a strong reason to use dlopen.

JanuszL replied:
I'm very biased so a second opinion is invaluable for me.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14718528480](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14718528480). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipe/recipe.yaml**:

 * ❌ There are 1 too many lines.  There should be one empty line at the end of the file.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/14721118557](https://github.com/conda-forge/conda-forge-webservices/actions/runs/14721118557). Examine the logs at this URL for more detail.</sub>

carterbox commented:
@JanuszL, could you please review / try some of these builds (you can download them from the build artifacts page on Azure e.g. https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1238164&view=artifacts&pathAsName=false&type=publishedArtifacts).

I was trying to run some of the upstream tests locally, but I couldn't figure out what to run or where to get the input files for some of the test.

JanuszL commented:
> @JanuszL, could you please review / try some of these builds (you can download them from the build artifacts page on Azure e.g. https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1238164&view=artifacts&pathAsName=false&type=publishedArtifacts).
> 
> I was trying to run some of the upstream tests locally, but I couldn't figure out what to run or where to get the input files for some of the test.

Will test them and get back to you (probably mid week).

isuruf commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,146 @@
+        - cudatoolkit
+    - if: match(cuda_compiler_version, "12.*")
+      then:
+        - cuda-cudart-static
```
How is the license requirements satisfied when this library is statically linked?

carterbox replied:
It's not.

- [ ] Add NVIDIA license or dynamically link cudart

carterbox commented:
@JanuszL, thanks for testing!

> By smoke test, I mean running a few test cases until failure, due to missing functionality. In this configuration, DALI is built with some features disabled, and our tests are designed to check a fully-featured DALI, so they not handle this configuration correctly.

What features are missing? We should open an issue to track those (e.g. missing dependencies, conda technical limitations, etc).

> However, the aarch64 package contains backend_impl.cpython-312-x86_64-linux-gnu.so, indicating it is likely broken - so not tests :(.

Probably CMake isn't able to correctly detect that we are cross-compiling. I'll have to look into this...

JanuszL commented:
@carterbox 
> @JanuszL, thanks for testing!
> 
> > By smoke test, I mean running a few test cases until failure, due to missing functionality. In this configuration, DALI is built with some features disabled, and our tests are designed to check a fully-featured DALI, so they not handle this configuration correctly.
> 
> What features are missing? We should open an issue to track those (e.g. missing dependencies, conda technical limitations, etc).

There are options that are on for full DALI (we can work on this in separate PRs as it may add too much to this one):
- BUILD_CVCUDA (CV-CUDA is a DALI submodule, it should not add more dependencies)
- BUILD_FFMPEG (we need to add FFmpeg to dependncies)
- BUILD_LIBSND (we need to add libsndfile)
- BUILD_LIBTAR (we need to add libtar)
- BUILD_LMDB (we need to add lmdb)
- BUILD_NVDEC (we need to build with BUILD_FFMPEG )
- BUILD_NVML (no additional dependency is involved)
- BUILD_NVOF (no additional dependency is involved)
- BUILD_NVTX (no additional dependency is involved)
- BUILD_NVCOMP (we need to add nvcomp)
- BUILD_AWSSDK (we need to add aws-sdk-cpp)

Should I create a ticket for each or an umbrela with a checklist?

carterbox commented:
Let's make a single issue in this feedstock with a checklist, please.

JanuszL commented:
> * BUILD_CVCUDA (CV-CUDA is a DALI submodule, it should not add more dependencies)
> * BUILD_FFMPEG (we need to add FFmpeg to dependncies)
> * BUILD_LIBSND (we need to add libsndfile)
> * BUILD_LIBTAR (we need to add libtar)
> * BUILD_LMDB (we need to add lmdb)
> * BUILD_NVDEC (we need to build with BUILD_FFMPEG )
> * BUILD_NVML (no additional dependency is involved)
> * BUILD_NVOF (no additional dependency is involved)
> * BUILD_NVTX (no additional dependency is involved)
> * BUILD_NVCOMP (we need to add nvcomp)
> * BUILD_AWSSDK (we need to add aws-sdk-cpp)

https://github.com/conda-forge/nvidia-dali-python-feedstock/issues/37

carterbox commented:
I'm pretty sure that the python extension modules are being compiled correctly for ARM, they are just named incorrectly by the custom functions. So, I just fix the name at the end of the build script.

I've also added a call to `file` on these binaries just as a sanity check that they claim to be compiled for the target architecture even if the name is wrong.

JanuszL commented:
> I'm pretty sure that the python extension modules are being compiled correctly for ARM, they are just named incorrectly by the custom functions. So, I just fix the name at the end of the build script.
> 
> I've also added a call to `file` on these binaries just as a sanity check that they claim to be compiled for the target architecture even if the name is wrong.

That is super weird because we ask python to give us the extension name - https://github.com/NVIDIA/DALI/blob/main/cmake/Utils.cmake#L346. When cross compiling we unfortunately need to abuse python-config - https://github.com/NVIDIA/DALI/blob/main/docker/Dockerfile.build.aarch64-linux#L112. So maybe this is where we land here.

JanuszL commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,145 @@
+    and inference applications.
+  license: Apache-2.0 AND BSD-3-Clause AND BSD-2-Clause
+  license_file:
+    - LICENSE
```
How about adding Acknowledgements.txt from DALI instead of all third_party/*? We aggregate all the necessary attributions there.

carterbox replied:
https://github.com/conda-forge/nvidia-dali-python-feedstock/pull/34/files#r2095939803

carterbox replied:
Thanks for the idea!

carterbox commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,145 @@
+    an execution engine for data processing to accelerate deep learning training
+    and inference applications.
+  license: Apache-2.0 AND BSD-3-Clause AND BSD-2-Clause
+  license_file:
+    - LICENSE
+    - COPYRIGHT
+    - third_party/README.rst
+    - third_party/cocoapi/license.txt
+    - third_party/ffts/COPYRIGHT
```
```suggestion
  license_file:
    - LICENSE
    - COPYRIGHT
    - Acknowledgements.txt  # Contains licenses for linked libraries
    - third_party/README.rst
```

JanuszL commented:
@carterbox 
I get in one of the tests:
``` dlopen libnvimgcodec.so failed!. Please install nvimagecodec: See https://developer.nvidia.com/nvimgcodec-downloads or simply do `pip install nvidia-nvimgcodec-cu12`.```
I see the lib is placed in:
```
/root/miniconda3/pkgs/nvidia-nvimagecodec-cuda120-0.5.0-dali_nvimagecodec/opt/nvidia/nvimgcodec/lib64/libnvimgcodec.so
/root/miniconda3/envs/conda_py310_env/opt/nvidia/nvimgcodec/lib64/libnvimgcodec.so
```

JanuszL commented:
Also for aarch64 it doesn't install libnvimgcodec as a dependency.

carterbox commented:
@JanuszL, I've added a patch to the dlopen failure messages to make it clearer what users should do if DALI cannot find an optional dependency. PTAL.

Also, you have to install dependencies from conda-forge. `$PREFIX/opt/nvidia/nvimgcodec/lib64/libnvimgcodec.so` is not the install location of the libnvimgcodec conda package, so I don't know why you have that binary there. It should be installed as `$PREFIX/lib/libnimgcodec.so.0`

> Also for aarch64 it doesn't install libnvimgcodec as a dependency.

nvimgcodec is optional; Do you mean that the libnvimgcodec package is not installable on linux-aarch64? I have also added a check for this in the test section. We will see when the new build is done whether that is the case.

JanuszL commented:
> @JanuszL, I've added a patch to the dlopen failure messages to make it clearer what users should do if DALI cannot find an optional dependency. PTAL.
> 
> Also, you have to install dependencies from conda-forge. `$PREFIX/opt/nvidia/nvimgcodec/lib64/libnvimgcodec.so` is not the install location of the libnvimgcodec conda package, so I don't know why you have that binary there. It should be installed as `$PREFIX/lib/libnimgcodec.so.0`
> 
> > Also for aarch64 it doesn't install libnvimgcodec as a dependency.
> 
> nvimgcodec is optional; Do you mean that the libnvimgcodec package is not installable on linux-aarch64? I have also added a check for this in the test section. We will see when the new build is done whether that is the case.

Ok, it could be a leftover from the test env I used. Let me recheck this again after installing libnvimgcodec directly.

JanuszL commented:
@carterbox I rechecked and installing libnvimgcodec helps. I think we are good.

conda-forge-admin commented:
Hi! This is the friendly conda-forge automerge bot!

I considered the following status checks when analyzing this PR:
 - **linter**: passed
 - **azure**: failed


Thus the PR was not passing and not merged.

carterbox commented on recipe/recipe.yaml:
```diff
@@ -0,0 +1,152 @@
+      run:
+        - pip
+        # Test that optional dependencies are co-installable
+        - libcufft
+        - libcufile
+        - libnpp
+        - libnvjpeg
```
```suggestion
        - if: match(cuda_compiler_version, "11.*")
          then:
            - cudatoolkit
        - if: match(cuda_compiler_version, "12.*")
          then:
            - libcufft
            - libcufile
            - libnpp
            - libnvjpeg
```
